src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+    } catch (ex) {
+        _marshal_exception_to_cs(data, ex);
+        return InvokeJSResult.Exception;
+    }
+}
+
+// C# signature would be `static TRes Invoke<T1, TRes>(ref T1 t1)` or `static void Invoke<T1>(ref T1 t1)`
+export function mono_wasm_invoke_js_1(interned_function_id: MonoString, signature: VoidPtr, data: VoidPtr): InvokeJSResult {
+    //console.log("mono_wasm_invoke_js_1 A " + signature + " " + data);
+    //console.log("mono_wasm_invoke_js_1 S0 " + getU32(<any>signature + (0 << 2)));
+    //console.log("mono_wasm_invoke_js_1 S1 " + getU32(<any>signature + (1 << 2)));
+    //console.log("mono_wasm_invoke_js_1 D0 " + getU32(<any>data + (0 << 2)));
+    //console.log("mono_wasm_invoke_js_1 D1 " + getU32(<any>data + (1 << 2)));
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        //console.log("mono_wasm_invoke_js_1 B");
+        const arg1_js = covert_arg_to_js(signature, data, 1);
+        //console.log("mono_wasm_invoke_js_1 C");
+        const res_js = fn(arg1_js);
+        //console.log("mono_wasm_invoke_js_1 D");
+        covert_result_to_cs(signature, data, res_js);
+        //console.log("mono_wasm_invoke_js_1 E");
+        return InvokeJSResult.Success;
+    } catch (ex) {
+        //console.log("mono_wasm_invoke_js_1 F");
+        _marshal_exception_to_cs(data, ex);
+        return InvokeJSResult.Exception;
+    }
+}
+
+export function mono_wasm_bind_js_function(interned_function_id: MonoString, arg_count: number, signature: VoidPtr, function_js_handle: Int32Ptr, is_exception: Int32Ptr): MonoObject {
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        const closure: any = { MONO, BINDING, INTERNAL, fn };
+        let body = "try {\n";
+
+        for (let position = 1; position <= arg_count; position++) {
+            const offset = position << 2;
+            const mono_type = <MonoType><any>getU32(<any>signature + offset);
+            const covertor = cs_to_js_map.get(mono_type);
+            assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
+            const convertor_name = `covertor${position}`;
+            closure[convertor_name] = covertor;
+
+            body += `  const arg${position}_js = closure.${convertor_name}(<any>data + ${offset});\n`;
+        }
+        if (arg_count === 0) {
+            body += "  const res_js = closure.fn();\n";
+        }
+        else if (arg_count === 1) {
+            body += "  const res_js = closure.fn(arg1_js);\n";
+        }
+
+        const res_mono_type = <MonoType><any>getU32(<any>signature);
+        if (res_mono_type === MonoTypeNull) {
+            body += "  assert(value === undefined, 'Function returned unexpected value, C# signature is void');\n";
+        } else {
+            const covertor = js_to_cs_map.get(res_mono_type);
+            assert(covertor, `Unknow covertor for type ${res_mono_type} at ret`);
+            closure["covertor0"] = covertor;
```

Comment:
If we always handle the 0 (result) slot specially, we should just call it the result converter instead of converter0

Reply:
done